### PR TITLE
Multiple Parameter Identifications Feature

### LINKED ISSUE #1336: Should be possible to run multiple Parameter Identifications in parallel in PK-Sim

### PR DESCRIPTION
**Pull request by abdelr**
Fixes #1336 (Support for multiple parameter identifications in parallel)

### CODE REVIEW

File: src/OSPSuite.Core/Domain/Services/ParameterIdentifications/ParameterIdentificationRunner.cs
Comment (abdelr) [reply src/OSPSuite.Core/Domain/Services/ParameterIdentifications/ParameterIdentificationRunner.cs]: I am not sure we will always have the same object for a PI I guess if at some point we clone or something but still keep the same ID (meaning it refers to the same PI) then we should identify the PI by its Id and not by the object itself. This is why I used Id and not the object.

File: src/OSPSuite.Presentation/Presenters/ParameterIdentifications/ParameterIdentificationFeedbackPresenter.cs
Comment (abdelr) [reply src/OSPSuite.Presentation/Presenters/ParameterIdentifications/ParameterIdentificationFeedbackPresenter.cs:60]: Again, to make sure copies (not sure if they exist) also works fine as long as they use the same Id. What is the need for an Id otherwise?

**Comment by abdelr** [comment]
Outdated code I will solve when one PR is merged, then the other will take the changes from develop. It is not clean if we keep merging changes from one PR into another. Makes it harder for you to review I think. So, when the first PR is accepted I will merge develop into the second and will solve conflicts. Sounds good?

File: src/OSPSuite.Core/Domain/Services/ParameterIdentifications/ParameterIdentificationRunner.cs
Comment (msevestre) [reply src/OSPSuite.Core/Domain/Services/ParameterIdentifications/ParameterIdentificationRunner.cs]: There should never be two objects with the same Id. It will break everything. 
Id are only used for
 - Serialization
 - Command so that we can retrieve an object by id without holding a reference 

Otherwise we always use reference. That's why I insist on using reference instead of id, especially to retrieve the same instance later on by id

Sometimes we store reference with "path" but this is for faster retrieval and this makes sense

File: src/OSPSuite.Presentation/Presenters/ParameterIdentifications/ParameterIdentificationFeedbackPresenter.cs
Comment (msevestre) [reply src/OSPSuite.Presentation/Presenters/ParameterIdentifications/ParameterIdentificationFeedbackPresenter.cs:60]: see above. Id are for internal use only. If you have reference, use them, There should never be two objects with the same id that are not the same. In fact, we overwritten equality in Entity to say that two objects are equal if they have the same id. Cloning an object should always reset ids

**Comment by msevestre** [comment]
>Outdated code I will solve when one PR is merged

I cannot merge the other one until you address my comment.s Specifically, removing the ConcurrencyManager is questionable for me as it was encapsulating a lot of the same code. THe implementation should use Parallel.ForEach but now we have lost a chance to make the code consistent everytime we want to run sthg in //. And we want to do more and more of this

**Comment by msevestre** [comment]
@abdelr How do you want to proceed here? I have remarks on both PR that we should at least discuss. I don't like having PR open for too long as it is a recipe for merge conflicts. Let me know

**Comment by abdelr** [comment]
> @abdelr How do you want to proceed here? I have remarks on both PR that we should at least discuss. I don't like having PR open for too long as it is a recipe for merge conflicts. Let me know

I think I have addressed now all the issues. We could discuss today in the meeting if something is missing. I would say this will be the feature branch (so, do not merge until the other PR is merged I would say) so I will update the other PR to merge into this branch.

File: src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs:71]: Hhere the results are crfeated by the caller. Why? Let's have the task return the results no?

File: src/OSPSuite.Core/Domain/Services/ParameterIdentifications/ParameterIdentificationRunner.cs
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Services/ParameterIdentifications/ParameterIdentificationRunner.cs:28]: you could use Any() instead of count. Does not matter here but it's more idiomatic...maybe

File: src/OSPSuite.R/Services/ConcurrentSimulationRunner.cs
Comment (msevestre) [inline src/OSPSuite.R/Services/ConcurrentSimulationRunner.cs:25]: not one one line.

File: src/OSPSuite.R/Services/ConcurrentSimulationRunner.cs
Comment (msevestre) [inline src/OSPSuite.R/Services/ConcurrentSimulationRunner.cs:125]: let's have the concurrentManager create the results so that you can always return

File: src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs
Comment (abdelr) [reply src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs]: I think there was a problem when throwing exceptions. If an exception is thrown then the whole list was lost. I rather think there was a bad exception handling so let us try again and see if it goes good with these changes.

File: src/OSPSuite.Core/Domain/Services/ParameterIdentifications/ParameterIdentificationRunner.cs
Comment (abdelr) [reply src/OSPSuite.Core/Domain/Services/ParameterIdentifications/ParameterIdentificationRunner.cs:28]: Sure, it is more idiomatic ;)

File: src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs:77]: Instnatiate that guy first. And then 
if(!data.any())
 return results;

**Review by msevestre — CHANGES_REQUESTED** [review]

File: src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs
Comment (abdelr) [reply src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs:107]: @msevestre No, the idea was to have the ability to overwrite the settings. So if you say for example the maximum number of cores in settings is 100 but for some reason you want to use less on a single call, you can still pass 50 on the function call. If you pass 0 (default) or lower it will use the value from settings. We could remove this argument though if you think it is overkilling. The settings can be change any time anyways so you could theoretically change the value in the settings, call the function and then restore the settings.

File: src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs
Comment (abdelr) [reply src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs:72]: It is actually not necessary. If there is no data, the foreach will not iterate meaning the return results at the end will return the empty list ;). I am removing those lines.

File: src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs
Comment (msevestre) [reply src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs:108]: Exactly. That's why using `_maximumNumberOfCoresToUse` is wrong because it's always this value and is set in the constructor
```
 _maximumNumberOfCoresToUse = Math.Max(1, coreUserSettings.MaximumNumberOfCoresToUse);

```
I think it should be numberOfCoreToUse or sthg like this

in fact the value `numberOfCoresToUse ` is only set but never used. Something is clearly off

Comment (msevestre) [reply src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs:108]: and while we are at it, numberOfCoresToUse =0 as default is very, very strange (or use a negative value??!?). Set it as null by default which would mean use the settings. Otherwise use this value no?

File: src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs
Comment (msevestre) [reply src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs]: yep.

File: src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs
Comment (msevestre) [reply src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs:108]: I think it's good to have this parameter however (probably for testing scenarios et.c.)

Comment (abdelr) [reply src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs]: You are right. It was used before in the function createParallelOptions but at some point was removed lol. I also changed it to uint so you are not allowed to pass negative values (which are meaningless in this context indeed)

**Review by msevestre — APPROVED** [review]
Looks good to me!
What's the next step? Do you want ot work on the feedback stuff?

File: src/OSPSuite.R/Services/ConcurrentSimulationRunner.cs
Comment (abdelr) [reply src/OSPSuite.R/Services/ConcurrentSimulationRunner.cs:160]: > Looks good to me! What's the next step? Do you want ot work on the feedback stuff?

I am planning to finish the UI feedback soon but first I am trying to debug an issue with the sensitivities calculations Vanessa has. So all in all yes, I will work on it but not right now.

**Comment by abdelr** [comment]
> Looks good to me! What's the next step? Do you want ot work on the feedback stuff?

I thought I reply to this message but it looks like I did not. I will work on the UI stuff but not right now. I wonder if it makes sense to merge this already to avoid conflicts and when I am finished with the issue I am addressing now I will go back to the feedback view (#1372 ). Sounds reasonable?

**Comment by msevestre** [comment]
ok going to merge
I jut added some remark on the issue mentioned above

https://github.com/Open-Systems-Pharmacology/OSPSuite.Core/issues/1372#issuecomment-967148656

Let's do it

**Pull request merged**
